sycl/doc/extensions/proposed/sycl_ext_oneapi_work_group_static.asciidoc

```diff
@@ -146,73 +143,89 @@ specified explicitly.
 usages of that function within the kernel. In cases where a function is called
 multiple times, developers must take care to avoid race conditions (e.g., by
 calling `group_barrier` before and after using the memory).
 ====
 
+Change to SYCL 2020 section `5.9.2 Common address space deduction rules`:
+Namespace scope: if the variable is `work_group_static` object,
+then the variable is assigned to the local address space.
+Otherwise normal rules applies.
+
 SYCL 2020 requires that all global variables accessed by a device function are
 `const` or `constexpr`. This extension lifts that restriction for
 `work_group_static` variables.
 
-[NOTE]
-====
-Since `work_group_static` acts as a view, wrapping an underlying pointer, a
-developer may still choose to declare variables as `const`.
-====
-
 When `T` is a class type or bounded array, the size of the allocation is known
-at compile-time, and a SYCL implementation may embed the size of the allocation
+at compile-time, and a SYCL implementation embeds the size of the allocation
 directly within a kernel. Each instance of `work_group_static<T>` is associated
 with a unique allocation in device local memory.
 
-When `T` is an unbounded array, the size of the allocation is unknown at
-compile-time, and must be communicated to the SYCL implementation via the
-`work_group_static_memory_size` property. Every instance of `work_group_static`
-for which `T` is an unbounded array is associated with a single, shared,
-allocation in device local memory. For example, two instances declared
-as `work_group_static<int[]>` and `work_group_static<float[]>` will be
-associated with the same shared allocation.
-
-If the total amount of device local memory requested (i.e., the sum of
-all memory requested by `local_accessor`, `group_local_memory`,
-`group_local_memory_for_overwrite` and `work_group_static`) exceeds a device's
-local memory capacity (as reported by `local_mem_size`) then the implementation
-must throw a synchronous `exception` with the `errc::memory_allocation` error
-code from the kernel invocation command (e.g. `parallel_for`).
-
 [source,c++]
 ----
-operator T&() const noexcept;
+operator T&() noexcept;
 ----
 _Returns_: A reference to the object stored in the device local memory
 associated with this instance of `work_group_static`.
 
 [source,c++]
 ----
-const work_group_static<T>& operator=(const T& value) const noexcept;
+work_group_static<T>& operator=(const T& value) noexcept;
 ----
-_Constraints_: Available only if `std::is_array_v<T>>` is false.
+_Constraints_: Available only if `std::is_array_v<T>` is false.
 
 _Effects_: Replaces the value referenced by `*ptr` with `value`.
 
 _Returns_: A reference to this instance of `work_group_static`.
 
 [source,c++]
 ----
-T* operator&() const noexcept;
+T* operator&() noexcept;
 ----
 _Returns_: A pointer to the device local memory associated with this
 instance of `work_group_static` (i.e., `ptr`).
 
+=== `get_dynamic_work_group_memory` function
```

Comment:
I'm confused, is this PR about `static` or `dynamic`? If both, can we split it in two?

Reply:
I think `static` refers to the fact that `work_group_static ` is implicitly a static / thread local object.
But that's a question for @Pennycook to answer. Maybe a split would be good TBH.

Reply:
Originally we'd talked about calling this `work_group_local` or `work_group_specific`, with support for both static (i.e. compile-time sized) and dynamic (i.e. run-time sized) allocations.  We later decided to split this into `work_group_static` and `work_group_dynamic`, but they're still grouped together here.

@gmlueck, does that match your recollection?  If the new name(s) are confusing, maybe it's worth revisiting some of the previous names?

If we went back to something like `work_group_specific` or `work_group_local` we could still highlight the difference in static/dynamic semantics via usage:

```c++
// store values directly in work-group-specific variables
/* optional: static */ syclex::work_group_specific<int> scalar; 
/* optional: static */ syclex::work_group_specific<int[16]> compile_time_array;

// store a shared pointer in a work-group specific variable
/* optional: static */ syclex::work_group_specific<int*> run_time_pointer;

// each work-item has a private pointer to the dynamic work-group memory
// NB: proposing to call this "scratch" (or something else) to highlight that only one such allocation exists
int* ptr = syclex::get_work_group_scratch_memory<int>();
```

Reply:
@Pennycook: This PR makes some significant changes to the API.  Are you OK with these?  I think there are two major changes:

* The class is no longer described as a "view".  Consequently, the `const` behavior is different.
* The `work_group_static` class no longer allows `T` to be an unbounded array.  Instead, this PR adds a namespace scope function named `get_dynamic_work_group_memory`.

I have a question about this second change.  We need some SYCLomatic migration plan for CUDA code that declares a namespace scope `__shared__` variable that is an unbounded array:

```
__shared__ float myglob[];

__global__ mykernel() {
  myglob[i] = /*...*/;
}
```

You cannot call `get_dynamic_work_group_memory` at namespace scope.  Would we instead replace every reference to `myglob` like this?

```
void mysyclkernel() {
  syclex::get_dynamic_work_group_memory<float>()[i] = /*...*/;
}
```

FWIW, we discussed this extension briefly at the Khronos F2F.  There was some skepticism about promoting `work_group_static` to KHR if `T` could be an unbounded array because the semantics are weird with the aliasing rules.  Therefore, it probably does make sense to move the unbounded array part to a new API.  In fact, we might even want to move it to a new extension.  That would allow us to promote all of `work_group_static` to a KHR.

> @gmlueck, does that match your recollection? If the new name(s) are confusing, maybe it's worth revisiting some of the previous names?

Getting around to this question now ... if we remove the support for unbounded arrays from `work_group_static`, then the name really does make sense.  All uses of `work_group_static` would allocate a statically-known amount of work-group memory.

Reply:
> @Pennycook: This PR makes some significant changes to the API

Thanks for pointing this out.  I didn't realize that the PR was also changing the specification.

> - The class is no longer described as a "view". Consequently, the const behavior is different.
> - The work_group_static class no longer allows T to be an unbounded array. Instead, this PR adds a namespace scope function named get_dynamic_work_group_memory.

I think this is okay.  The two main reasons for designing this as a view initially were compatibility with the SYCL 2020 rules about what can be captured by kernels, and a need to support the use-case where multiple objects could represent the same (dynamic) memory allocation.  The first issue can be worked around by lifting the restriction, and the second issue is solved by introducing a separate `get_` function for the dynamic case.

> You cannot call get_dynamic_work_group_memory at namespace scope. Would we instead replace every reference to myglob like this?

I think calling `get_dynamic()` for every use is pretty nasty.  One option would be to introduce something like a `work_item_specific`/`work_item_static` -- the combination of a per-work-item pointer at namespace scope and a scratchpad gives you what you want:

```c++
syclex::work_item_specific<float*> myglob;

void mysyclkernel() {
  // would need to generate this initialization upon kernel entry
  myglob = syclex::get_work_group_scratch_memory<float>();
  
  // all subsequent uses of myglob (anywhere) would use the scratchpad
  myglob[i] = /*...*/;
}
```

> There was some skepticism about promoting work_group_static to KHR if T could be an unbounded array because the semantics are weird with the aliasing rules.

I agree that the aliasing rules for unbound arrays were weird.  But if we switch to a `get_scratch()` function then I think there's a good chance we could still go to a KHR.  It's clearly fetching one thing, so the aliasing isn't a surprise.

> Getting around to this question now ... if we remove the support for unbounded arrays from work_group_static, then the name really does make sense. All uses of work_group_static would allocate a statically-known amount of work-group memory.

That's true.  I do think, though, that we should be careful how we describe this -- without a `work_group_dynamic` it's less obvious that the `static` means "statically known amount" and not the same as C++ `static`.  It's especially confusing because `work_group_static` variables _are_ implicitly C++ `static`.

Maybe we just need to not draw attention to the size aspect, and just say that they're called `work_group_static` because they're C++ per-work-group and always behave as if the `static` keyword was specified?  Then the confusion is gone.